pkg/services/dashboards/database/database.go

```diff
@@ -847,12 +847,22 @@ func (d *dashboardStore) deleteAlertDefinition(dashboardId int64, sess *db.Sessi
 func (d *dashboardStore) GetDashboard(ctx context.Context, query *dashboards.GetDashboardQuery) (*dashboards.Dashboard, error) {
 	var queryResult *dashboards.Dashboard
 	err := d.store.WithDbSession(ctx, func(sess *db.Session) error {
-		if query.ID == 0 && len(query.Slug) == 0 && len(query.UID) == 0 {
+		if query.ID == 0 && len(query.UID) == 0 && (query.Title == nil || query.FolderID == nil) {
 			return dashboards.ErrDashboardIdentifierNotSet
 		}
 
-		dashboard := dashboards.Dashboard{Slug: query.Slug, OrgID: query.OrgID, ID: query.ID, UID: query.UID}
-		has, err := sess.Get(&dashboard)
+		dashboard := dashboards.Dashboard{OrgID: query.OrgID, ID: query.ID, UID: query.UID}
+		mustCols := []string{}
+		if query.Title != nil {
+			dashboard.Title = *query.Title
+			mustCols = append(mustCols, "title")
+		}
+		if query.FolderID != nil {
+			dashboard.FolderID = *query.FolderID
+			mustCols = append(mustCols, "folder_id")
+		}
```

Comment:
if it's empty then the `UQE_dashboard_org_id_folder_id_title` index won't be used
maybe it's not a big problem given that another index (`IDX_dashboard_title`) is used, but still has to filter out rows using where (for the `org_id`)

```
mysql> EXPLAIN SELECT * FROM dashboard WHERE org_id = 1 AND title = 'gdev dashboards';
+----+-------------+-----------+------------+------+---------------------------------------------------------------------------------------------------------------------------------------+---------------------+---------+-------+------+----------+-------------+
| id | select_type | table     | partitions | type | possible_keys                                                                                                                         | key                 | key_len | ref   | rows | filtered | Extra       |
+----+-------------+-----------+------------+------+---------------------------------------------------------------------------------------------------------------------------------------+---------------------+---------+-------+------+----------+-------------+
|  1 | SIMPLE      | dashboard | NULL       | ref  | UQE_dashboard_org_id_folder_id_title,UQE_dashboard_org_id_uid,IDX_dashboard_org_id,IDX_dashboard_org_id_plugin_id,IDX_dashboard_title | IDX_dashboard_title | 758     | const |    1 |   100.00 | Using where |
+----+-------------+-----------+------------+------+---------------------------------------------------------------------------------------------------------------------------------------+---------------------+---------+-------+------+----------+-------------+
1 row in set, 1 warning (0.01 sec)

mysql> EXPLAIN SELECT * FROM dashboard WHERE org_id = 1 AND title = 'gdev dashboards' AND folder_id = 0;
+----+-------------+-----------+------------+-------+---------------------------------------------------------------------------------------------------------------------------------------+--------------------------------------+---------+-------------------+------+----------+-------+
| id | select_type | table     | partitions | type  | possible_keys                                                                                                                         | key                                  | key_len | ref               | rows | filtered | Extra |
+----+-------------+-----------+------------+-------+---------------------------------------------------------------------------------------------------------------------------------------+--------------------------------------+---------+-------------------+------+----------+-------+
|  1 | SIMPLE      | dashboard | NULL       | const | UQE_dashboard_org_id_folder_id_title,UQE_dashboard_org_id_uid,IDX_dashboard_org_id,IDX_dashboard_org_id_plugin_id,IDX_dashboard_title | UQE_dashboard_org_id_folder_id_title | 774     | const,const,const |    1 |   100.00 | NULL  |
+----+-------------+-----------+------------+-------+---------------------------------------------------------------------------------------------------------------------------------------+--------------------------------------+---------+-------------------+------+----------+-------+
1 row in set, 1 warning (0.01 sec)
```

Reply:
If it’s nil, then either UID or ID is set and those have indices?

Reply:
not sure: the following condition:
https://github.com/grafana/grafana/blob/882b0f608ac0ebf38d0727050a3e980a19763d34/pkg/services/dashboards/database/database.go#L850
allows only the Title to be set, no?

I think that it has to become: `query.Title == nil && query.FolderID == nil)` instead

Reply:
my mistake, I think that you are right